File: backends/dotnet/services/Geo/Geo.Client/CQRS/Handlers/C/SetOnDisk.cs
// -----------------------------------------------------------------------
// <copyright file="SetOnDisk.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace D2.Geo.Client.CQRS.Handlers.C;

using System.Net;
using D2.Shared.Handler;
using D2.Shared.I18n;
using D2.Shared.Result;
using D2.Shared.Utilities.Constants;
using Google.Protobuf;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;
using H = D2.Geo.Client.Interfaces.CQRS.Handlers.C.ICommands.ISetOnDiskHandler;
using I = D2.Geo.Client.Interfaces.CQRS.Handlers.C.ICommands.SetOnDiskInput;
using O = D2.Geo.Client.Interfaces.CQRS.Handlers.C.ICommands.SetOnDiskOutput;

/// <summary>
/// Handler for setting georeference data on disk.
/// </summary>
public partial class SetOnDisk : BaseHandler<SetOnDisk, I, O>, H
{
    private readonly string r_filePath;

    /// <summary>
    /// Initializes a new instance of the <see cref="SetOnDisk"/> class.
    /// </summary>
    ///
    /// <param name="config">
    /// The configuration.
    /// </param>
    /// <param name="context">
    /// The handler context.
    /// </param>
    public SetOnDisk(
        IConfiguration config,
        IHandlerContext context)
        : base(context)
    {
        var dataDir = config[Constants.LOCAL_FILES_PATH_CONFIG_KEY] ?? "./data";
        Directory.CreateDirectory(dataDir);
        r_filePath = Path.Combine(dataDir, Constants.GEO_REF_DATA_FILE_NAME);
    }

    /// <inheritdoc />
    protected override HandlerOptions DefaultOptions => new(LogInput: false, LogOutput: false);

    /// <summary>
    /// Executes the handler to set georeference data on disk.
    /// </summary>
    ///
    /// <param name="input">
    /// The input parameters for the handler.
    /// </param>
    /// <param name="ct">
    /// The cancellation token.
    /// </param>
    ///
    /// <returns>
    /// The result of the set operation.
    /// </returns>
    protected override async ValueTask<D2Result<O?>> ExecuteAsync(
        I input,
        CancellationToken ct = default)
    {
        try
        {
            var bytes = input.Data.ToByteArray();
            await File.WriteAllBytesAsync(r_filePath, bytes, ct);

            return D2Result<O?>.Ok(new O());
        }
        catch (IOException ex)
        {
            LogDiskWriteFailed(Context.Logger, ex, TraceId);

            return D2Result<O?>.Fail(
                [TK.Geo.Errors.DISK_WRITE_FAILED],
                HttpStatusCode.InternalServerError);
        }

        // Let the base handler catch any other exceptions.
    }

    /// <summary>
    /// Logs an error when an IOException occurs while writing georeference data to disk.
    /// </summary>
    [LoggerMessage(EventId = 1, Level = LogLevel.Error, Message = "IOException occurred while writing georeference data to disk. TraceId: {TraceId}")]
    private static partial void LogDiskWriteFailed(ILogger logger, Exception ex, string? traceId);
}

File: backends/node/services/geo/geo-client/src/handlers/c/set-on-disk.ts
import { mkdir, writeFile } from "node:fs/promises";
import { join } from "node:path";
import { BaseHandler, type IHandlerContext } from "@d2/handler";
import { D2Result } from "@d2/result";
import { GeoRefData as GeoRefDataCodec } from "@d2/protos";
import { GEO_REF_DATA_FILE_NAME } from "@d2/utilities";
import type { GeoClientOptions } from "../../geo-client-options.js";
import { Commands } from "../../interfaces/index.js";

type Input = Commands.SetOnDiskInput;
type Output = Commands.SetOnDiskOutput;

/**
 * Handler for persisting georeference data to disk as protobuf binary.
 * Mirrors D2.Geo.Client.CQRS.Handlers.C.SetOnDisk in .NET.
 */
export class SetOnDisk extends BaseHandler<Input, Output> implements Commands.ISetOnDiskHandler {
  override get redaction() {
    return Commands.SET_ON_DISK_REDACTION;
  }

  private readonly filePath: string;

  constructor(options: GeoClientOptions, context: IHandlerContext) {
    super(context);
    this.filePath = join(options.dataDir, GEO_REF_DATA_FILE_NAME);
  }

  protected async executeAsync(input: Input): Promise<D2Result<Output | undefined>> {
    try {
      const bytes = GeoRefDataCodec.encode(input.data).finish();
      await mkdir(join(this.filePath, ".."), { recursive: true });
      await writeFile(this.filePath, bytes);
      return D2Result.ok({ data: {} });
    } catch (err: unknown) {
      this.context.logger.error(
        `IOException occurred while writing georeference data to disk. TraceId: ${this.traceId}`,
        { error: err instanceof Error ? err.message : String(err) },
      );
      return D2Result.unhandledException({
        messages: ["Unable to write to disk."],
      });
    }
  }
}

export type { SetOnDiskInput, SetOnDiskOutput } from "../../interfaces/c/set-on-disk.js";
